pkg/resourcemanager/disk/reconcile.go

```diff
@@ -0,0 +1,207 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+package disk
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	azurev1alpha1 "github.com/Azure/azure-service-operator/api/v1alpha1"
+	"github.com/Azure/azure-service-operator/pkg/errhelp"
+	"github.com/Azure/azure-service-operator/pkg/helpers"
+	"github.com/Azure/azure-service-operator/pkg/resourcemanager"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+func (g *AzureDiskClient) Ensure(ctx context.Context, obj runtime.Object, opts ...resourcemanager.ConfigOption) (bool, error) {
+
+	instance, err := g.convert(obj)
+	if err != nil {
+		return true, err
+	}
+
+	client := getDiskClient()
+
+	location := instance.Spec.Location
+	resourceGroup := instance.Spec.ResourceGroup
+	resourceName := instance.Name
+	createOption := instance.Spec.CreateOption
+	diskSizeGB := instance.Spec.DiskSizeGB
+
+	instance.Status.Provisioning = true
+
+	// Check if this item already exists. This is required
+	// to overcome the issue with the lack of idempotence of the Create call
+	item, err := g.GetDisk(ctx, resourceGroup, resourceName)
+	if err == nil {
+		instance.Status.Provisioned = true
+		instance.Status.Provisioning = false
+		instance.Status.Message = resourcemanager.SuccessMsg
+		instance.Status.ResourceId = *item.ID
+		return true, nil
+	}
+
+	// TODO: Update this validation as support for new CreateOption types is added.
+	// Validate CreateOption. Currently only 'Empty' is supported.
+	if !strings.EqualFold(createOption, "Empty") {
+		instance.Status.Provisioned = false
+		instance.Status.Provisioning = false
+		instance.Status.Message = errhelp.UnsupportedDiskCreateOption
+		return false, nil
+	}
+
+	future, err := g.CreateDisk(
```

Comment:
If you already called create disk once you can use the `item.ProvisioningState` from the previous "GetDisk" call to tell when it is done provisioning so you can avoid calling create multiple times.